Move ProcessThrottler to AuxiliaryProcessProxy

pvollan opened the pull request:
#### b3da4ad472fe86eed672da39d2a33da5b05fcdc9
<pre>
Move ProcessThrottler to AuxiliaryProcessProxy
<a href="https://bugs.webkit.org/show_bug.cgi?id=269136">https://bugs.webkit.org/show_bug.cgi?id=269136</a>
<a href="https://rdar.apple.com/122708925">rdar://122708925</a>

Reviewed by Chris Dumez.

We can simplify the code by moving the ProcessThrottler object from every subclass to the
AuxiliaryProcessProxy base class.

This was previously landed in &lt;<a href="https://commits.webkit.org/275195@main">https://commits.webkit.org/275195@main</a>&gt; and later reverted
because it introduced a crash. The reason for the crash was that the ProcessThrottler
was being deleted later since it moved to AuxiliaryProcessProxy. That could result in
pure virtual function calls in the AuxiliaryProcessProxy destructor, since the
ProcessThrottler calls virtual methods in AuxiliaryProcessProxy. This patch addresses
this crash by calling didDisconnectFromProcess in AuxiliaryProcessProxy::shutDownProcess.

* Source/WebKit/UIProcess/AuxiliaryProcessProxy.h:
(WebKit::AuxiliaryProcessProxy::sendProcessDidResume):
(WebKit::AuxiliaryProcessProxy::clientName const):
* Source/WebKit/UIProcess/AuxiliaryProcessProxy.cpp:
(WebKit::AuxiliaryProcessProxy::didFinishLaunching):
* Source/WebKit/UIProcess/AuxiliaryProcessProxy.h:
(WebKit::AuxiliaryProcessProxy::throttler):
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
(WebKit::GPUProcessProxy::didFinishLaunching):
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.h:
* Source/WebKit/UIProcess/Model/ModelProcessProxy.cpp:
(WebKit::ModelProcessProxy::didFinishLaunching):
* Source/WebKit/UIProcess/Model/ModelProcessProxy.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::didFinishLaunching):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::didFinishLaunching):
* Source/WebKit/UIProcess/WebProcessProxy.h:
(WebKit::WebProcessProxy::throttler const): Deleted.

Canonical link: <a href="https://commits.webkit.org/275497@main">https://commits.webkit.org/275497@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3966a2f582792b1180cbfb783d180f1608184819

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/41950 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/20968 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/44343 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/44536 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/38048 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/24129 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/18296 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/34767 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/42524 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/17891 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/36126 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/15700 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/15572 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/45941 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/38135 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/37484 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/41385 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/16766 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/13774 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/39776 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/18385 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/18444 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5637 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/18030 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/UIProcess/AuxiliaryProcessProxy.h:
```diff
@@ -214,10 +214,10 @@ class AuxiliaryProcessProxy
+    virtual void sendProcessDidResume(ResumeReason) { };
     virtual void didChangeThrottleState(ProcessThrottleState) { };
-    virtual ASCIILiteral clientName() const = 0;
+    virtual ASCIILiteral clientName() const { return nullptr; }
```
Why is this change needed?

cdumez replied:
If you did this as a fix for your earlier crashes, I do not believe this is correct. It is not OK to call virtual functions in constructors and destructors. It doesn't matter if pure virtual or not.

pvollan replied:
Ah, good point! I will fix in the upcoming patch.

Thanks for reviewing!

cdumez commented on Source/WebKit/UIProcess/AuxiliaryProcessProxy.h:
```diff
@@ -214,10 +214,10 @@ class AuxiliaryProcessProxy
 #endif
 
     enum ResumeReason : bool { ForegroundActivity, BackgroundActivity };
-    virtual void sendPrepareToSuspend(IsSuspensionImminent, double remainingRunTime, CompletionHandler<void()>&&) = 0;
```
Why is this change needed?

cdumez commented on Source/WebKit/UIProcess/AuxiliaryProcessProxy.h:
```diff
@@ -214,10 +214,10 @@ class AuxiliaryProcessProxy
     enum ResumeReason : bool { ForegroundActivity, BackgroundActivity };
-    virtual void sendPrepareToSuspend(IsSuspensionImminent, double remainingRunTime, CompletionHandler<void()>&&) = 0;
-    virtual void sendProcessDidResume(ResumeReason) = 0;
+    virtual void sendPrepareToSuspend(IsSuspensionImminent, double remainingRunTime, CompletionHandler<void()>&&) { };
```
Why is this change needed?

cdumez commented on Source/WebKit/UIProcess/AuxiliaryProcessProxy.h:
```diff
@@ -214,10 +214,10 @@ class AuxiliaryProcessProxy
-    virtual void sendPrepareToSuspend(IsSuspensionImminent, double remainingRunTime, CompletionHandler<void()>&&) = 0;
-    virtual void sendProcessDidResume(ResumeReason) = 0;
+    virtual void sendPrepareToSuspend(IsSuspensionImminent, double remainingRunTime, CompletionHandler<void()>&&) { };
+    virtual void sendProcessDidResume(ResumeReason) { };
```
Why is this change needed?

cdumez commented on Source/WebKit/UIProcess/AuxiliaryProcessProxy.h:
```diff
@@ -67,7 +67,7 @@ class AuxiliaryProcessProxy
 
 protected:
-    AuxiliaryProcessProxy(bool alwaysRunsAtBackgroundPriority = false, Seconds responsivenessTimeout = ResponsivenessTimer::defaultResponsivenessTimeout);
+    AuxiliaryProcessProxy(bool shouldTakeUIBackgroundAssertion, bool alwaysRunsAtBackgroundPriority = false, Seconds responsivenessTimeout = ResponsivenessTimer::defaultResponsivenessTimeout);
```
Those 2 first boolean parameters seem a bit errors prone. Trivial for a call site to mix / reverse the 2. We should probably switch to enum classes.

cdumez commented on Source/WebKit/UIProcess/AuxiliaryProcessProxy.cpp:
```diff
@@ -378,6 +390,7 @@ void AuxiliaryProcessProxy::replyToPendingMessages()
 
 void AuxiliaryProcessProxy::shutDownProcess()
 {
+    fprintf(stderr, "shutting down process");
```
Likely didn't mean to keep this.

cdumez commented on Source/WebKit/UIProcess/AuxiliaryProcessProxy.cpp:
```diff
@@ -77,6 +78,8 @@ AuxiliaryProcessProxy::AuxiliaryProcessProxy(bool alwaysRunsAtBackgroundPriority
 
 AuxiliaryProcessProxy::~AuxiliaryProcessProxy()
 {
+    m_throttler.invalidateAllActivitiesAndDropAssertion();
```
I'm still not convinced this is safe, as I assume this can end up calling virtual functions on the AuxiliaryProcessProxy. Maybe we should set a flag in ~ProcessThrottler() to keep track of the fact that we started destruction and not call the bad virtual function in this case? I assume the one that's causing trouble is calling prepareToSuspend() on destruction when releasing the assertion?

pvollan replied:
Ah, good point! As you say, this is also not correct, since we can still be calling virtual methods in AuxiliaryProcessProxy from ~AuxiliaryProcessProxy. The issue is that `ProcessThrottler::sendPrepareToSuspendIPC`, which has several AuxiliaryProcessProxy virtual method calls can be called from ~AuxiliaryProcessProxy. As you suggest, a flag is a good approach to solve this. I am also looking into a patch which calls ProcessThrottler::didDisconnectFromProcess from AuxiliaryProcessProxy::shutDownProcess.

Thanks for reviewing!

cdumez commented on Source/WebKit/UIProcess/AuxiliaryProcessProxy.cpp:
```diff
@@ -66,9 +66,10 @@ static Seconds adjustedTimeoutForThermalState(Seconds timeout)
+AuxiliaryProcessProxy::AuxiliaryProcessProxy(ShouldTakeUIBackgroundAssertion shouldTakeUIBackgroundAssertion, AlwaysRunsAtBackgroundPriority alwaysRunsAtBackgroundPriority, Seconds responsivenessTimeout)
     : m_responsivenessTimer(*this, adjustedTimeoutForThermalState(responsivenessTimeout))
-    , m_alwaysRunsAtBackgroundPriority(alwaysRunsAtBackgroundPriority)
+    , m_alwaysRunsAtBackgroundPriority(alwaysRunsAtBackgroundPriority == AlwaysRunsAtBackgroundPriority::Yes ? true : false)
```
Should just be `alwaysRunsAtBackgroundPriority == AlwaysRunsAtBackgroundPriority::Yes`. No need for a ternary.

cdumez commented on Source/WebKit/UIProcess/AuxiliaryProcessProxy.cpp:
```diff
@@ -66,9 +66,10 @@ static Seconds adjustedTimeoutForThermalState(Seconds timeout)
     : m_responsivenessTimer(*this, adjustedTimeoutForThermalState(responsivenessTimeout))
-    , m_alwaysRunsAtBackgroundPriority(alwaysRunsAtBackgroundPriority)
+    , m_alwaysRunsAtBackgroundPriority(alwaysRunsAtBackgroundPriority == AlwaysRunsAtBackgroundPriority::Yes ? true : false)
+    , m_throttler(*this, shouldTakeUIBackgroundAssertion == ShouldTakeUIBackgroundAssertion::Yes ? true : false)
```
Ditto.

cdumez commented on Source/WebKit/UIProcess/AuxiliaryProcessProxy.h:
```diff
@@ -222,6 +225,8 @@ class AuxiliaryProcessProxy
     virtual void prepareToDropLastAssertion(CompletionHandler<void()>&& completionHandler) { completionHandler(); }
     virtual void didDropLastAssertion() { }
 
+    bool isBeingDeleted() const { return m_isBeingDeleted; }
```
This can be private. We don't even need a getter.

cdumez commented on Source/WebKit/UIProcess/AuxiliaryProcessProxy.cpp:
```diff
@@ -378,6 +390,8 @@ void AuxiliaryProcessProxy::replyToPendingMessages()
 
 void AuxiliaryProcessProxy::shutDownProcess()
 {
+    throttler().didDisconnectFromProcess();
```
Could we do this in a ScopeExit? This will release the assertion. However, `platformStartConnectionTerminationWatchdog()` below will take an assertion again. It'd be nice if the process didn't suspend and resume because we release the assertion and take a new one right after. We can do better in terms of ordering here.

cdumez commented on Source/WebKit/UIProcess/ProcessThrottler.cpp:
```diff
@@ -394,6 +394,8 @@ void ProcessThrottler::sendPrepareToSuspendIPC(IsSuspensionImminent isSuspension
     } else {
         m_pendingRequestToSuspendID = generatePrepareToSuspendRequestID();
         Ref process = m_process.get();
+        if (process->isBeingDeleted())
```
Have you considered relying on the `m_isConnectedToProcess` flag instead?

pvollan replied:
That is a good point, will fix!

Thanks for reviewing!

cdumez reviewed (approved):
LGTM

pvollan committed:
Move ProcessThrottler to AuxiliaryProcessProxy
https://bugs.webkit.org/show_bug.cgi?id=269136
rdar://122708925

Reviewed by Chris Dumez.

We can simplify the code by moving the ProcessThrottler object from every subclass to the
AuxiliaryProcessProxy base class.

This was previously landed in <https://commits.webkit.org/275195@main> and later reverted
because it introduced a crash. The reason for the crash was that the ProcessThrottler
was being deleted later since it moved to AuxiliaryProcessProxy. That could result in
pure virtual function calls in the AuxiliaryProcessProxy destructor, since the
ProcessThrottler calls virtual methods in AuxiliaryProcessProxy. This patch addresses
this crash by calling didDisconnectFromProcess in AuxiliaryProcessProxy::shutDownProcess.

* Source/WebKit/UIProcess/AuxiliaryProcessProxy.h:
(WebKit::AuxiliaryProcessProxy::sendProcessDidResume):
(WebKit::AuxiliaryProcessProxy::clientName const):
* Source/WebKit/UIProcess/AuxiliaryProcessProxy.cpp:
(WebKit::AuxiliaryProcessProxy::didFinishLaunching):
* Source/WebKit/UIProcess/AuxiliaryProcessProxy.h:
(WebKit::AuxiliaryProcessProxy::throttler):
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.cpp:
(WebKit::GPUProcessProxy::didFinishLaunching):
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.h:
* Source/WebKit/UIProcess/Model/ModelProcessProxy.cpp:
(WebKit::ModelProcessProxy::didFinishLaunching):
* Source/WebKit/UIProcess/Model/ModelProcessProxy.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::didFinishLaunching):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::didFinishLaunching):
* Source/WebKit/UIProcess/WebProcessProxy.h:
(WebKit::WebProcessProxy::throttler const): Deleted.

Canonical link: https://commits.webkit.org/275497@main
